infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectFactory.java

```diff
@@ -103,6 +117,33 @@ public OpenShiftProject create(String workspaceId) throws InfrastructureExceptio
     return osProject;
   }
 
+  @Override
+  public void delete(String workspaceId) throws InfrastructureException {
+    String projectName;
+    try {
+      projectName = evalNamespaceName(workspaceId, EnvironmentContext.getCurrent().getSubject());
+    } catch (NotFoundException | ServerException | ConflictException | ValidationException e) {
+      throw new InfrastructureException(
+          format("Failed to evaluate the project name to use for workspace %s.", workspaceId), e);
+    }
+
+    OpenShiftProject osProject = doCreateProject(workspaceId, projectName);
+    osProject.delete();
+  }
+
+  @VisibleForTesting
+  @Override
+  protected String evalNamespaceName(String workspaceId, Subject currentUser)
+      throws NotFoundException, ServerException, InfrastructureException, ConflictException,
+          ValidationException {
+    return super.evalNamespaceName(workspaceId, currentUser);
+  }
+
+  @Override
+  protected boolean checkNamespaceExists(String namespaceName) throws InfrastructureException {
```

Comment:
I'd argue that is not a great name either, because it could be understood that a non-existent namespace is also available, because it can/will be created.

What is the actual difference between a namespace that doesn't exist and a namespace that the server cannot see due to RBAC? If the namespace exists but the user can't access it, we're failing either sooner or later.

Reply:
We can imagine some method like
```
  if (checkNamespaceExists(namespaceName)) {
    project.delete();
  }
```
and in case of RBAC permissions issue it would be better to have something like
```
  if (checkNamespaceExists(namespaceName)) {
    project.delete();
  } else {
    throw new InfrastructureException("the current user is not permitted to remove namespace");
  }
```

> I'd argue that is not a great name either, because it could be understood that a non-existent namespace is also available, because it can/will be created.

Java docs can explain that `available` - exists and user is permitted to use it. Or another word can be chosen.

Reply:
```
 if (checkNamespaceExists(namespaceName)) {
    project.delete();
  } else {
    throw new InfrastructureException("the current user is not permitted to remove namespace");
  }
```

You get very similar behavior from just attempting the delete, don't you?